code/game/machinery/doors/brig_system.dm

```diff
@@ -9,350 +15,347 @@
 	var/list/obj/item/paper/incident/incident_reports = list()
-	var/obj/item/paper/incident/current_report = null
 
-	var/current_menu = "main_menu"
+	/// A list of World objects this machine controls.
+	var/list/obj/structure/machinery/targets = list()
+
+	/// The report currently ticking down the clock.
+	var/obj/item/paper/incident/active_report = null
+
+	/// The report currently being viewed.
+	var/obj/item/paper/incident/viewed_report = null
 
 	maptext_height = 26
 	maptext_width = 32
 
-/obj/structure/machinery/brig_cell/Initialize()
-	. = ..()
-	addtimer(CALLBACK(src, .proc/search_for_components), 20)
+/obj/structure/machinery/brig_cell/attack_hand(mob/user)
+	if(..() || !allowed(usr) || inoperable())
+		return
 
-/obj/structure/machinery/brig_cell/proc/search_for_components()
-	for(var/obj/structure/machinery/door/window/brigdoor/M in machines)
-		if(M.id == id)
-			targets += M
+	tgui_interact(user)
 
-	for(var/obj/structure/machinery/flasher/F in machines)
-		if(F.id == id)
-			targets += F
+/obj/structure/machinery/brig_cell/tgui_interact(mob/user, datum/tgui/ui)
+	ui = SStgui.try_update_ui(user, src, ui)
+	if (!ui)
+		ui = new(user, src, "BrigCell", "Brig Cell")
+		ui.open()
 
-	for(var/i in GLOB.brig_locker_list)
-		var/obj/structure/closet/secure_closet/brig/C = i
-		if(C.id == id)
-			targets += C
+/obj/structure/machinery/brig_cell/ui_data(mob/user)
+	var/list/data = list()
 
-	for(var/obj/structure/machinery/door/poddoor/almayer/locked/P in machines)
-		if(P.id == id)
-			targets += P
+	data["viewing_incident"] = viewed_report ? TRUE : FALSE
 
-	if(length(targets) == 0)
-		stat |= BROKEN
+	if (viewed_report)
+		var/datum/crime_incident/incident = viewed_report.incident
 
-	update_icon()
+		var/time_left = get_time_left(incident)
 
-/obj/structure/machinery/brig_cell/process()
-	if(inoperable())
-		return
+		data["suspect"] = incident.criminal_name
+		data["progress"] = 1 - (time_left / 60) / incident.brig_sentence
+		data["sentence"] = (incident.status & BRIG_SENTENCE_PERMA) ? "Permabrig" : "[incident.brig_sentence] Minutes"
+		data["time_left"] = "[num2text((time_left / 60) % 60)]:[add_zero(num2text(time_left % 60), 2)]"
 
-	if(!incident_reports)
-		stop_processing()
+		// Flags.
+		data["active"]	= (incident.status & BRIG_SENTENCE_ACTIVE) // Is this timer currently ticking?
+		data["started"]	= (incident.time_served || data["active"]) // Has this timer been started at all?
 
-	for(var/obj/item/paper/incident/I in incident_reports)
-		var/datum/crime_incident/incident = I.incident
+		if (incident.status & BRIG_SENTENCE_PARDONED)
+			data["status_text"] = "PARDONED"
+			data["status_class"] = "info"
+		else if (incident.status & BRIG_SENTENCE_PERMA)
+			data["status_text"] = "PERMABRIG"
+			data["status_class"] = "danger"
+		else if (incident.status & BRIG_SENTENCE_SERVED)
+			data["status_text"] = "SERVED"
+			data["status_class"] = "success"
+		else
+			data["status_text"] = null
+			data["status_class"] = null
```

Comment:
why not just send the incident_status to the UI and calculate these all in-UI? look at the escape pod tgui for implementation example